sdk/core/System.ClientModel/src/Options/LoggingOptions.cs

```diff
@@ -0,0 +1,177 @@
+    private IList<string> _loggedQueryParameters = s_defaultLoggedQueryParameters;
+    private string? _clientAssembly;
+    private string? _requestIdHeaderName;
+
+    /// <summary>
+    /// Get or sets value indicating whether HTTP pipeline logging is enabled.
+    /// </summary>
+    public bool IsLoggingEnabled
+    {
+        get => _isLoggingEnabled;
+        set
+        {
+            AssertNotFrozen();
+
+            _isLoggingEnabled = value;
+        }
+    }
+
+    /// <summary>
+    /// Gets or sets value indicating if request or response content should be logged.
+    /// </summary>
+    public bool IsLoggingContentEnabled
+    {
+        get => _isLoggingContentEnabled;
+        set
+        {
+            AssertNotFrozen();
+
+            _isLoggingContentEnabled = value;
+        }
+    }
+
+    /// <summary>
+    /// Gets or sets value indicating maximum size of content to log in bytes. Defaults to 4096.
+    /// </summary>
+    public int LoggedContentSizeLimit
+    {
+        get => _loggedContentSizeLimit;
+        set
+        {
+            AssertNotFrozen();
+
+            _loggedContentSizeLimit = value;
+        }
+    }
+
+    /// <summary>
+    /// Gets or sets a list of header names that are not redacted during logging.
+    /// </summary>
+    public IList<string> LoggedHeaderNames
+    {
+        get => _loggedHeaderNames;
+        set
+        {
+            AssertNotFrozen();
+
+            _loggedHeaderNames = value;
+        }
+    }
+
+    /// <summary>
+    /// Gets or sets a list of query parameter names that are not redacted during logging.
+    /// </summary>
+    public IList<string> LoggedQueryParameters
+    {
+        get => _loggedQueryParameters;
+        set
+        {
+            AssertNotFrozen();
+
+            _loggedQueryParameters = value;
+        }
+    }
+
+    /// <summary>
+    /// Gets or sets the name of the client assembly to include in logging.
+    /// </summary>
+    public string? LoggedClientAssemblyName
+    {
+        get => _clientAssembly;
+        set
+        {
+            AssertNotFrozen();
+
+            _clientAssembly = value;
+        }
+    }
+
+    /// <summary>
+    /// Gets or sets the header name that contains the request Id to include in logging.
+    /// </summary>
+    public string? RequestIdHeaderName
+    {
+        get => _requestIdHeaderName;
+        set
+        {
+            AssertNotFrozen();
+
+            _requestIdHeaderName = value;
+        }
+    }
+
+    /// <summary>
+    /// Freeze this instance of <see cref="LoggingOptions"/>.  After this method
+    /// has been called, any attempt to set properties on the instance or call
+    /// methods that would change its state will throw <see cref="InvalidOperationException"/>.
+    /// </summary>
+    public virtual void Freeze()
+    {
+        _frozen = true;
+        _loggedHeaderNames = new ReadOnlyCollection<string>(_loggedHeaderNames.ToArray());
```

Comment:
Is ToArray needed here?

Reply:
If I remove the setters on these properties I don't think it would be needed anymore. I was doing this to prevent changes in the underlying collection from being reflected: https://learn.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.readonlycollection-1?view=net-8.0#remarks

Reply:
Looking again do you think we should guard against this scenario?
```csharp
LoggingOptions myOptions = new LoggingOptions();
var headerNames = myOptions.AllowedHeaderNames;

// create pipeline

headerNames.Add("Header");
```